src/test/codegen/option-nonzero-eq.rs

```diff
@@ -0,0 +1,34 @@
+// compile-flags: -O -Zmerge-functions=disabled
+
+#![crate_type = "lib"]
+
+extern crate core;
+use core::num::{NonZeroU32, NonZeroI64};
+use core::ptr::NonNull;
+
+// CHECK-lABEL: @non_zero_eq
+#[no_mangle]
+pub fn non_zero_eq(l: Option<NonZeroU32>, r: Option<NonZeroU32>) -> bool {
+    // CHECK: start:
+    // CHECK-NEXT: %2 = icmp eq i32 %0, %1
+    // CHECK-NEXT: ret i1 %2
```

Comment:
Up to you: these test are currently coupled to the names that LLVM happens to pick for the variables while optimizing.  That's probably fine, but in case different LLVM versions make different naming/ordering choices, you might consider a simpler filecheck like
```rust
    // CHECK: start:
    // CHECK-NEXT: icmp eq i32
    // CHECK-NEXT: ret i1
```
As that's still enough to test that it's exactly one comparison then a return, which I think is the core of the point of the test.  If it was `icmp eq i32 %1, %0`, that would be ok too, for example, as would `icmp eq i32 %l, %r`.  And presumably there are Rust functionality tests that make sure it's not `icmp eq i32 %0, %0` -- not that LLVM would ever leave that in, as if that mistake happened then the IR would just be `start: ret i1 true`.

There's also the option of something like
```rust
    // CHECK: start:
    // CHECK-NEXT: %[[CMP:.+]] = icmp eq i32
    // CHECK-NEXT: ret i1 %[[CMP]]
```
if you wanted to be extra-sure what it's returning, but with the `CHECK-NEXT` I don't think there's anything else it could plausibly return.  With only the three lines, there's really no other possibility.